### PR TITLE
MGDAPI-1905 Implement a workaround for the lack of STS support in 3Scale

### PR DESCRIPTION
**Pull request by valerymo**
# PR deprecated, please see new PR: https://github.com/integr8ly/integreatly-operator/pull/2837
# Issue link
https://issues.redhat.com/browse/MGDAPI-1905

# What
3Scale does not support STS mode of authentication with AWS, and we need to provide a temporary workaround.
Our workaround would be to ask customer to create an IAM account dedicated to 3Scale and pass credentials for this IAM account through two new addon parameters. This IAM account would allow access only to S3 bucket(s) with a predefined name(s). 
This logic is implemented in current PR

# Verification steps
**STS Cluster**
- Prepare STS cluster
- Install Rhoam from this branch
- Open your cluster OSD UI
- Open Secret *addon-managed-api-service-parameters* in Project *redhat-rhoam-operator*
  - Edit Secret, set your values for *tmp-s3-access-key-id* and *tmp-s3-secret-access-key* parameters
- Open Secret *s3-credentials* in Project *redhat-rhoam-3scale*
  - Reveal values and check that *AWS_ACCESS_KEY_ID* and *AWS_SECRET_ACCESS_KEY* - updated, and have same values as *tmp-s3-access-key-id* and *tmp-s3-secret-access-key* parameters in *addon-managed-api-service-parameters* in Project *redhat-rhoam-operator*.

**NON-STS cluster**
- Prepare Non-STS cluster
- Do similar steps as for STS cluster
- See that *AWS_ACCESS_KEY_ID* and *AWS_SECRET_ACCESS_KEY* parameters of *s3-credentials* Secret have not changed.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -28,6 +28,8 @@ SMTP_PASS ?= ''
 SMTP_PORT ?= ''
 SMTP_FROM ?= ''
 ROLE_ARN ?= "arn:aws:iam::485026278258:role/12345"
+TMP_S3_ACCESS_KEY_ID ?= "123"
```
Comment (laurafitzgerald) [inline Makefile:31]: Could remove the use of `TMP_` from the env vars and yaml keys.

Comment (valerymo) [reply Makefile]: Done. Thank you

File: pkg/products/threescale/reconciler.go [hunk 2]
```diff
@@ -900,14 +902,44 @@ func (r *Reconciler) getBlobStorageFileStorageSpec(ctx context.Context, serverCl
...
-				credSec.Data["AWS_SECRET_ACCESS_KEY"] = blobStorageSec.Data["credentialSecretKey"]
+				if isSTS {
+					credentialKeyID, found, err := addon.GetStringParameter(context.TODO(), serverClient, addOnSecretNamespace, sts.CredsTmpS3SecretAccessKey)
+					if err != nil {
+						return fmt.Errorf("Error getting addon parameter %s: %w", sts.CredsTmpS3SecretAccessKey, err)
+					}
+					if !found {
+						return fmt.Errorf("AddOn parameter %s not found in secret %s", sts.CredsTmpS3SecretAccessKey, addon.DefaultSecretName)
+					}
+					credSec.Data["AWS_SECRET_ACCESS_KEY"] = []byte(credentialKeyID)
```
Comment (KevFan) [inline pkg/products/threescale/reconciler.go:939]:
```suggestion
				if isSTS {
					credentialSecretKey, found, err := addon.GetStringParameter(context.TODO(), serverClient, addOnSecretNamespace, sts.CredsTmpS3SecretAccessKey)
					if err != nil {
						return fmt.Errorf("Error getting addon parameter %s: %w", sts.CredsTmpS3SecretAccessKey, err)
					}
					if !found {
						return fmt.Errorf("AddOn parameter %s not found in secret %s", sts.CredsTmpS3SecretAccessKey, addon.DefaultSecretName)
					}
					credSec.Data["AWS_SECRET_ACCESS_KEY"] = []byte(credentialSecretKey)
```

Using `credentialSecretKey` has the variable name probably would be better :thinking:

File: pkg/resources/sts/sts.go [hunk 3]
```diff
@@ -13,19 +13,21 @@ import (
...
+	ClusterCloudCredentialName  = "cluster"
 	RoleArnParameterName        = "sts-role-arn"
 	RoleSessionName             = "Red-Hat-cloud-resources-operator"
 	CredsSecretName             = "sts-credentials"
 	CredsSecretRoleARNKeyName   = "role_arn"
 	CredsSecretTokenPathKeyName = "web_identity_token_file"
 	CredsRoleEnvKey             = "ROLE_ARN"
 	CredsTokenPathEnvKey        = "TOKEN_PATH"
+	CredsTmpS3AccessKeyId       = "s3-access-key-id"
+	CredsTmpS3SecretAccessKey   = "s3-secret-access-key"
```
Comment (KevFan) [inline pkg/resources/sts/sts.go:25]:
```suggestion
	CredsS3AccessKeyId       = "s3-access-key-id"
	CredsS3SecretAccessKey   = "s3-secret-access-key"
```
Aligning to remove `Tmp` from name :thinking:

File: pkg/resources/sts/sts.go [hunk 4]
```diff
@@ -85,3 +87,38 @@ func GetSTSCredentialsFromEnvVar() (string, string, error) {
 
 	return roleARN, tokenPath, nil
 }
+
+// GetS3BucketCredentials retrieves Credential addon parameters to be used by 3scale
+// Using S3 Bucket Credentiol addon parameters is temporary solution,
+// workaround for the lack of STS support in 3Scale,
+// as noted in https://issues.redhat.com/browse/MGDAPI-1905,
+// until our feature request https://issues.redhat.com/browse/THREESCALE-7132 gets implemented
+func GetS3BucketCredentials(ctx context.Context, client k8sclient.Client, namespace string) (string, string, error) {
```
Comment (KevFan) [inline pkg/resources/sts/sts.go:96]: This function doesn't seem used, are we keeping it for some reason ? :thinking:

**Review by KevFan — COMMENTED** [review]
Looks good overall, left some comments :technologist:

File: pkg/products/threescale/reconciler.go [hunk 5]
```diff
@@ -900,14 +902,44 @@ func (r *Reconciler) getBlobStorageFileStorageSpec(ctx context.Context, serverCl
 		Data: map[string][]byte{},
 	}
 
+	r.log.Info("checking if STS mode")
+	isSTS, err := sts.IsClusterSTS(ctx, serverClient, r.log)
+	if err != nil {
+		return nil, fmt.Errorf("Error checking STS mode: %w", err)
+	}
+	addOnSecretNamespace := "redhat-rhoam-operator"
```
[the remark this answers is not in the document]
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: Done. Thank you

File: pkg/products/threescale/reconciler.go (hunk 2 above)
Comment (valerymo) [reply pkg/products/threescale/reconciler.go]: Done. Thank you

File: pkg/resources/sts/sts.go (hunk 3 above)
Comment (valerymo) [reply pkg/resources/sts/sts.go]: Done. Thank you

File: pkg/resources/sts/sts.go (hunk 4 above)
Comment (valerymo) [reply pkg/resources/sts/sts.go]: Done. Thank you very much for review and comments

File: pkg/products/threescale/reconciler.go [hunk 6]
```diff
@@ -900,14 +902,44 @@ func (r *Reconciler) getBlobStorageFileStorageSpec(ctx context.Context, serverCl
...
 			case "credentialSecretKey":
-				credSec.Data["AWS_SECRET_ACCESS_KEY"] = blobStorageSec.Data["credentialSecretKey"]
+				if isSTS {
+					credentialSecretKey, found, err := addon.GetStringParameter(context.TODO(), serverClient, addOnSecretNamespace, sts.CredsS3SecretAccessKey)
+					if err != nil {
+						return fmt.Errorf("Error getting addon parameter %s: %w", sts.CredsS3SecretAccessKey, err)
+					}
+					if !found {
+						return fmt.Errorf("AddOn parameter %s not found in secret %s", sts.CredsS3SecretAccessKey, addon.DefaultSecretName)
+					}
```
Comment (KevFan) [inline pkg/products/threescale/reconciler.go:938]: Seems there a slight decrease in code coverage due to not unit testing condition blocks where there is an error or if the addon parameter field is not found in addon secret, would be nice if we can try unit testing these also :thinking:

**Comment by KevFan** [comment]
* Verified `s3-credentials` secret is using fields from Blobstorage secret on non sts cluster
* Verified `s3-credentials` secret is using fields from addon parametes secret on sts cluster

File: pkg/products/threescale/reconciler.go (hunk 6 above)
Comment (KevFan) [reply pkg/products/threescale/reconciler.go:938]: @valerymo If you think the above is not feasible, then I think this is good to merge as is, I'll put an `lgtm` if that is the case

**Comment by KevFan** [comment]
Closing, superceded by https://github.com/integr8ly/integreatly-operator/pull/2837

**Pull request closed without merging**